File: algopt/rebalancer/benchmarks/interface_benchmarks/DynamicDimensionBenchmark.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "algopt/rebalancer/benchmarks/interface_benchmarks/BenchmarkUtils.h"
#include "algopt/rebalancer/interface/ProblemSolver.h"

#include <folly/Benchmark.h>
#include <folly/container/irange.h>
#include <folly/init/Init.h>

#include <string_view>

namespace facebook::rebalancer::interface::benchmarks {
namespace {
constexpr std::string_view DIMENSION_NAME = "test_dimension";
constexpr std::string_view SCOPE_NAME = "test_scope";
constexpr std::string_view PARTITION_NAME = "test_partition";

struct ProblemData {
  std::unique_ptr<ProblemSolver> solver;
  folly::F14FastMap<std::string, folly::F14FastMap<std::string, double>>
      dimension;
};

struct ProblemDataWithPartition {
  std::unique_ptr<ProblemSolver> solver;
  folly::F14FastMap<std::string, folly::F14FastMap<std::string, double>>
      scopeItemToGroupToValue;
  folly::F14FastMap<std::string, std::vector<std::string>> groupToObjects;
};

void addScopeToSolver(ProblemSolver& solver, int nScopeItems) {
  folly::F14FastMap<std::string, std::string> containerToScopeItem;
  for (const auto scopeIdx : folly::irange(nScopeItems)) {
    containerToScopeItem[fmt::format("host_{}", scopeIdx)] =
        fmt::format("scope_{}", scopeIdx);
  }
  solver.addScope(std::string(SCOPE_NAME), containerToScopeItem);
}

} // namespace

std::unique_ptr<ProblemData> setupProblemAndGetDynamicDimension(
    int nScopeItems,
    int nObjects,
    bool canExecuteAsync = false) {
  auto solver = createBaseProblemSolver(nScopeItems, nObjects, canExecuteAsync);
  addScopeToSolver(*solver, nScopeItems);

  auto dimension = createDynamicDimensionData(nScopeItems, nObjects);

  return std::make_unique<ProblemData>(ProblemData{
      .solver = std::move(solver), .dimension = std::move(dimension)});
}

std::unique_ptr<ProblemDataWithPartition>
setupProblemAndGetDynamicDimensionByPartition(
    int nScopeItems,
    int nGroups,
    int nObjects,
    bool canExecuteAsync = false) {
  auto solver = createBaseProblemSolver(nScopeItems, nObjects, canExecuteAsync);
  addScopeToSolver(*solver, nScopeItems);

  auto groupToObjects = createPartitionData(nGroups, nObjects);
  solver->addPartition(std::string(PARTITION_NAME), groupToObjects);

  auto scopeItemToGroupToValue = createDynamicDimensionData(
      nScopeItems, nGroups, /*dimensionBasedOnGroups=*/true);

  return std::make_unique<ProblemDataWithPartition>(ProblemDataWithPartition{
      .solver = std::move(solver),
      .scopeItemToGroupToValue = std::move(scopeItemToGroupToValue),
      .groupToObjects = std::move(groupToObjects)});
}

BENCHMARK(AddDynamicDimension_100x90k) {
  folly::BenchmarkSuspender suspend;
  auto data = setupProblemAndGetDynamicDimension(/*nScopeItems=*/100,
                                                 /*nObjects=*/90000);
  suspend.dismiss();

  data->solver->addDynamicObjectDimension(
      std::string(DIMENSION_NAME),
      std::string(SCOPE_NAME),
      data->dimension,
      0.0);

  BENCHMARK_SUSPEND {
    data.reset();
  }
}

BENCHMARK(AddDynamicDimension_250x2m) {
  folly::BenchmarkSuspender suspend;
  auto data = setupProblemAndGetDynamicDimension(/*nScopeItems=*/250,
                                                 /*nObjects=*/2000000);
  suspend.dismiss();

  data->solver->addDynamicObjectDimension(
      std::string(DIMENSION_NAME),
      std::string(SCOPE_NAME),
      data->dimension,
      0.0);

  BENCHMARK_SUSPEND {
    data.reset();
  }
}

BENCHMARK(AddDynamicDimensionByPartition) {
  folly::BenchmarkSuspender suspend;
  constexpr int nScopeItems = 500;
  constexpr int nGroups = 2000;
  constexpr int nObjects = nGroups * 500;
  auto data = setupProblemAndGetDynamicDimensionByPartition(
      nScopeItems, nGroups, nObjects);
  suspend.dismiss();

  data->solver->addDynamicObjectDimension(
      std::string(DIMENSION_NAME),
      std::string(SCOPE_NAME),
      std::string(PARTITION_NAME),
      data->scopeItemToGroupToValue,
      0.0);

  BENCHMARK_SUSPEND {
    data.reset();
  }
}

BENCHMARK(AddDynamicDimensionByPartition_250x2m_with_1_partition) {
  folly::BenchmarkSuspender suspend;
  constexpr int nScopeItems = 250;
  constexpr int nGroups = 1;
  constexpr int nObjects = 2000000;
  auto data = setupProblemAndGetDynamicDimensionByPartition(
      nScopeItems, nGroups, nObjects);
  suspend.dismiss();

  data->solver->addDynamicObjectDimension(
      std::string(DIMENSION_NAME),
      std::string(SCOPE_NAME),
      std::string(PARTITION_NAME),
      data->scopeItemToGroupToValue,
      0.0);

  BENCHMARK_SUSPEND {
    data.reset();
  }
}

BENCHMARK(AddDynamicDimensionByPartitionSparse) {
  folly::BenchmarkSuspender suspend;
  constexpr int nScopeItems = 500;
  constexpr int nGroups = 2000;
  constexpr int nObjects = nGroups * 500;
  constexpr int groupsToInclude = 10;

  auto data = setupProblemAndGetDynamicDimensionByPartition(
      nScopeItems, nGroups, nObjects);

  folly::F14FastMap<std::string, folly::F14FastMap<std::string, double>>
      sparseData;
  sparseData.reserve(nScopeItems);
  for (const auto& [scopeItem, groupToValue] : data->scopeItemToGroupToValue) {
    folly::F14FastMap<std::string, double> sparseGroupToValue;
    sparseGroupToValue.reserve(groupsToInclude);
    int count = 0;
    for (const auto& [group, value] : groupToValue) {
      if (count++ >= groupsToInclude) {
        break;
      }
      sparseGroupToValue[group] = value;
    }
    sparseData.emplace(scopeItem, std::move(sparseGroupToValue));
  }
  suspend.dismiss();

  data->solver->addDynamicObjectDimension(
      std::string(DIMENSION_NAME),
      std::string(SCOPE_NAME),
      std::string(PARTITION_NAME),
      sparseData,
      0.0);

  BENCHMARK_SUSPEND {
    data.reset();
  }
}

void runSeveralAddDynamicDimension_400x360kAsync(
    bool canExecuteAsync,
    bool includeTimeToDestruct) {
  folly::BenchmarkSuspender suspend;
  constexpr int kNumDimensions = 3;
  constexpr int nScopeItems = 400;
  constexpr int nObjects = 360000;
  auto solver = createBaseProblemSolver(nScopeItems, nObjects, canExecuteAsync);
  addScopeToSolver(*solver, nScopeItems);
  suspend.dismiss();

  for (const auto dimensionIdx : folly::irange(kNumDimensions)) {
    folly::BenchmarkSuspender innerSuspend;
    auto dimension = createDynamicDimensionData(nScopeItems, nObjects);
    innerSuspend.dismiss();

    solver->addDynamicObjectDimension(
        fmt::format("{}_{}", DIMENSION_NAME, dimensionIdx),
        std::string(SCOPE_NAME),
        std::move(dimension),
        0.0);
  }
  if (includeTimeToDestruct) {
    solver.reset();
  } else {
    BENCHMARK_SUSPEND {
      solver.reset();
    }
  }
}

BENCHMARK(Several400x360k_NotAsync) {
  runSeveralAddDynamicDimension_400x360kAsync(
      /*canExecuteAsync=*/false,
      /*includeTimeToDestruct=*/false);
}

BENCHMARK_RELATIVE(Several400x360k_Async_waitForDestructor) {
  runSeveralAddDynamicDimension_400x360kAsync(/*canExecuteAsync=*/true,
                                              /*includeTimeToDestruct=*/true);
}

BENCHMARK_RELATIVE(Several400x360k_Async) {
  runSeveralAddDynamicDimension_400x360kAsync(
      /*canExecuteAsync=*/true,
      /*includeTimeToDestruct=*/false);
}

} // namespace facebook::rebalancer::interface::benchmarks

int main(int argc, char** argv) {
  const folly::Init init(&argc, &argv);
  folly::runBenchmarks();
  return 0;
}

File: algopt/rebalancer/interface/py_client/benchmarks/interface_benchmarks/dynamic_dimension_benchmark.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# pyre-strict
from dataclasses import dataclass

from algopt.rebalancer.interface.py_client.benchmarks.interface_benchmarks.benchmark_utils import (
    create_base_problem_solver,
    create_dynamic_dimension_data,
    create_scope_mapping,
)
from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from windtunnel.benchmarks.python_benchmark_runner.benchmark import (
    main,
    register_benchmark,
)


FILE_PATH: str = (
    "algopt/rebalancer/interface/py_client/benchmarks/dynamic_dimension_benchmark.py"
)

DIMENSION_NAME = "test_dimension"
SCOPE_NAME = "test_scope"
PARTITION_NAME = "test_partition"


@dataclass
class ProblemData:
    solver: ProblemSolver
    dimension: dict[str, dict[str, float]]


def setup_problem(
    num_scope_items: int,
    num_objects: int,
    can_execute_async: bool = False,
) -> ProblemData:
    solver = create_base_problem_solver(num_scope_items, num_objects, can_execute_async)
    solver.addScope(SCOPE_NAME, create_scope_mapping(num_scope_items))
    return ProblemData(
        solver, create_dynamic_dimension_data(num_scope_items, num_objects)
    )


DATA_400x360k: ProblemData = setup_problem(num_scope_items=400, num_objects=360000)
DATA_400x360k_ASYNC: ProblemData = setup_problem(
    num_scope_items=400, num_objects=360000, can_execute_async=True
)
DATA_100x90k: ProblemData = setup_problem(num_scope_items=100, num_objects=90000)
DATA_250x2m: ProblemData = setup_problem(250, 2000000)


@register_benchmark(FILE_PATH)
def benchmark_add_dynamic_dimension_100x90k() -> None:
    DATA_100x90k.solver.addDynamicObjectDimension(
        dimensionName=DIMENSION_NAME,
        scope=SCOPE_NAME,
        scopeItemToObjectToValue=DATA_100x90k.dimension,
        defaultValue=0.0,
    )


@register_benchmark(FILE_PATH)
def benchmark_add_dynamic_dimension_250x2m() -> None:
    DATA_250x2m.solver.addDynamicObjectDimension(
        dimensionName=DIMENSION_NAME,
        scope=SCOPE_NAME,
        scopeItemToObjectToValue=DATA_250x2m.dimension,
        defaultValue=0.0,
    )


@register_benchmark(FILE_PATH)
def benchmark_add_dynamic_dimension_250x2m_with_partition() -> None:
    DATA_250x2m.solver.addDynamicObjectDimensionByPartition(
        dimensionName=DIMENSION_NAME,
        scope=SCOPE_NAME,
        partitionName=PARTITION_NAME,
        scopeItemToGroupToValue=DATA_250x2m.dimension,
        defaultValue=0.0,
    )


@register_benchmark(FILE_PATH)
def benchmark_several_add_dynamic_dimension_400x360k_not_async() -> None:
    dimensionCount: int = 3
    for i in range(dimensionCount):
        DATA_400x360k.solver.addDynamicObjectDimension(
            dimensionName=f"{DIMENSION_NAME}_{i}",
            scope=SCOPE_NAME,
            scopeItemToObjectToValue=DATA_400x360k.dimension,
            defaultValue=0.0,
        )


@register_benchmark(FILE_PATH)
def benchmark_several_add_dynamic_dimension_400x360k_async() -> None:
    dimensionCount: int = 3
    for i in range(dimensionCount):
        DATA_400x360k_ASYNC.solver.addDynamicObjectDimension(
            dimensionName=f"{DIMENSION_NAME}_{i}",
            scope=SCOPE_NAME,
            scopeItemToObjectToValue=DATA_400x360k_ASYNC.dimension,
            defaultValue=0.0,
        )


if __name__ == "__main__":
    main()
